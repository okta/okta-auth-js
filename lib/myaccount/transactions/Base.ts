import { OktaAuthHttpInterface } from '../../http/types';

export type TransactionLink = {
  href: string;
  hints?: {
    allow?: string[];
  };
}

export type TransactionLinks = {
  self: TransactionLink;
  [property: string]: TransactionLink;
}

type TransactionOptions = {
  // TODO: move res type to http module
  res: {
    headers: Record<string, string>;
    _links?: Record<string, TransactionLink>;
    [property: string]: unknown;
  };
  accessToken: string;
};

export default class BaseTransaction {
  // Deprecated
  headers?: Record<string, string>;

  constructor(oktaAuth: OktaAuthHttpInterface, options: TransactionOptions) {
    const { res } = options;
    const { headers, ...rest } = res;
    
    // assign required fields from res
    if (headers) {
      this.headers = headers;
    }

    // add all rest fields from res
    Object.keys(rest).forEach(key => {
      if (key === '_links') {
        return;
      }
      this[key] = rest[key];
    });
  }
}

export interface TransactionType<T extends BaseTransaction = BaseTransaction> extends Function {
  new (oktaAuth: OktaAuthHttpInterface, options: TransactionOptions): T;
  prototype: T;
}
