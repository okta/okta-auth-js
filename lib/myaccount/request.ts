import { 
  default as BaseTransaction,
  TransactionType,
  TransactionLinks
} from './transactions/Base';
import { httpRequest } from '../http';
import { AuthSdkError } from '../errors';
import { MyAccountRequestOptions as RequestOptions } from './types';
import { OktaAuthOAuthInterface } from '../oidc/types';

type SendRequestOptions = RequestOptions & {
  url: string;
  method: string;
}

/* eslint-disable complexity */
export async function sendRequest<
  T extends BaseTransaction = BaseTransaction,
  N extends 'plural' | 'single' = 'single',
  NT = N extends 'plural' ? T[] : T
> (
  oktaAuth: OktaAuthOAuthInterface, 
  options: SendRequestOptions,
  TransactionClass: TransactionType<T> = BaseTransaction as TransactionType<T>,
): Promise<NT> {
  const {
    accessToken: accessTokenObj
  } = oktaAuth.tokenManager.getTokensSync();
  
  const accessToken = options.accessToken || accessTokenObj?.accessToken;
  const issuer = oktaAuth.getIssuerOrigin();
  const { url, method, payload } = options;
  const requestUrl = url.startsWith(issuer!) ? url : `${issuer}${url}`;

  if (!accessToken) {
    throw new AuthSdkError('AccessToken is required to request MyAccount API endpoints.');
  }
  
  const res = await httpRequest(oktaAuth, {
    headers: { 'Accept': '*/*;okta-version=1.0.0' },
    accessToken,
    url: requestUrl,
    method,
    ...(payload && { args: payload })
  });

  let ret: T | T[];
  if (Array.isArray(res)) {
    ret = res.map(item => new TransactionClass(oktaAuth, { 
      res: item, 
      accessToken
    }));
  } else {
    ret = new TransactionClass(oktaAuth, { 
      res, 
      accessToken
    });
  }
  return ret as NT;
}
/* eslint-enable complexity */

export type GenerateRequestFnFromLinksOptions = {
  oktaAuth: OktaAuthOAuthInterface;
  accessToken: string;
  methodName: string;
  links: TransactionLinks;
}

type IRequestFnFromLinks<T extends BaseTransaction> = (payload?) => Promise<T>;

export function generateRequestFnFromLinks<T extends BaseTransaction>(
  {
    oktaAuth, 
    accessToken,
    methodName,
    links,
  }: GenerateRequestFnFromLinksOptions,
  TransactionClass: TransactionType<T> = BaseTransaction as TransactionType<T>,
): IRequestFnFromLinks<T> {
  for (const method of ['GET', 'POST', 'PUT', 'DELETE']) {
    if (method.toLowerCase() === methodName) {
      const link = links.self;
      return (async (payload?) => sendRequest<T, 'single'>(oktaAuth, {
        accessToken,
        url: link.href,
        method,
        payload,
      }, TransactionClass));
    }
  }
  
  const link = links[methodName];
  if (!link) {
    throw new AuthSdkError(`No link is found with methodName: ${methodName}`);
  }

  return (async (payload?) => sendRequest<T, 'single'>(oktaAuth, {
    accessToken,
    url: link.href,
    method: link.hints!.allow![0],
    payload,
  }, TransactionClass));
}
