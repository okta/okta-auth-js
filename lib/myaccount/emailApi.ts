import { sendRequest } from './request';
import { IAPIFunction } from './types';
import {
  BaseTransaction, 
  EmailTransaction, 
  EmailChallengeTransaction 
} from './transactions';

/**
 * @scope: okta.myAccount.email.read
 */
export const getEmails: IAPIFunction<EmailTransaction[]> = async (
  oktaAuth,
  options?
) => {
  const transaction = await sendRequest<EmailTransaction, 'plural'>(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'GET',
    accessToken: options?.accessToken
  }, EmailTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.email.read
 */
export const getEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'GET',
    accessToken,
  }, EmailTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.email.manage
 */
export const addEmail: IAPIFunction<EmailTransaction> = async (
  oktaAuth, 
  options
): Promise<EmailTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/emails',
    method: 'POST',
    payload,
    accessToken,
  }, EmailTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.email.manage
 */
export const deleteEmail: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}`,
    method: 'DELETE',
    accessToken
  });
  return transaction;
};

/**
 * @scope: okta.myAccount.email.read
 */
export const sendEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${id}/challenge`,
    method: 'POST',
    accessToken,
  }, EmailChallengeTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.email.read
 */
export const getEmailChallenge: IAPIFunction<EmailChallengeTransaction> = async (
  oktaAuth, 
  options
) => {
  const { emailId, challengeId, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}`,
    method: 'POST',
    accessToken,
  }, EmailChallengeTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.email.manage
 */
export const verifyEmailChallenge: IAPIFunction<BaseTransaction> = async (
  oktaAuth,
  options
) => {
  const { emailId, challengeId, payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/emails/${emailId}/challenge/${challengeId}/verify`,
    method: 'POST',
    payload,
    accessToken
  });
  return transaction;
};
