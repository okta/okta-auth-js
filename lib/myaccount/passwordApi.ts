import { sendRequest } from './request';
import { IAPIFunction } from './types';
import {
  BaseTransaction, 
  PasswordTransaction
} from './transactions';

/**
 * @scope: okta.myAccount.password.read
 */
export const getPassword: IAPIFunction<PasswordTransaction> = async (
  oktaAuth,
  options
) => {
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/password`,
    method: 'GET',
    accessToken: options?.accessToken,
  }, PasswordTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.password.manage
 */
export const enrollPassword: IAPIFunction<PasswordTransaction> = async (
  oktaAuth, 
  options
): Promise<PasswordTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/password',
    method: 'POST',
    payload,
    accessToken,
  }, PasswordTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.password.manage
 */
export const updatePassword: IAPIFunction<PasswordTransaction> = async (
  oktaAuth, 
  options
): Promise<PasswordTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/password',
    method: 'PUT',
    payload,
    accessToken,
  }, PasswordTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.password.manage
 */
export const deletePassword: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options?
) => {
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/password`,
    method: 'DELETE',
    accessToken: options?.accessToken,
  });
  return transaction;
};
