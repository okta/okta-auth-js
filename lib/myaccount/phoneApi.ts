import { sendRequest } from './request';
import { IAPIFunction } from './types';
import {
  BaseTransaction, 
  PhoneTransaction
} from './transactions';

/**
 * @scope: okta.myAccount.phone.read
 */
export const getPhones: IAPIFunction<PhoneTransaction[]> = async (
  oktaAuth,
  options?
) => {
  const transaction = await sendRequest<PhoneTransaction, 'plural'>(oktaAuth, {
    url: '/idp/myaccount/phones',
    method: 'GET',
    accessToken: options?.accessToken,
  }, PhoneTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.phone.read
 */
export const getPhone: IAPIFunction<PhoneTransaction> = async (
  oktaAuth,
  options
) => {
  const { accessToken, id } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/phones/${id}`,
    method: 'GET',
    accessToken,
  }, PhoneTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.phone.manage
 */
export const addPhone: IAPIFunction<PhoneTransaction> = async (
  oktaAuth, 
  options
): Promise<PhoneTransaction> => {
  const { accessToken, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/phones',
    method: 'POST',
    payload,
    accessToken,
  }, PhoneTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.phone.manage
 */
export const deletePhone: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options
) => {
  const { id, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/phones/${id}`,
    method: 'DELETE',
    accessToken,
  });
  return transaction;
};

/**
 * @scope: okta.myAccount.phone.manage
 */
export const sendPhoneChallenge: IAPIFunction<BaseTransaction> = async (
  oktaAuth, 
  options
) => {
  const { accessToken, id, payload } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/phones/${id}/challenge`,
    method: 'POST',
    payload,
    accessToken
  });
  return transaction;
};

/**
 * @scope: okta.myAccount.phone.manage
 */
export const verifyPhoneChallenge: IAPIFunction<BaseTransaction> = async (
  oktaAuth,
  options
) => {
  const { id, payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: `/idp/myaccount/phones/${id}/verify`,
    method: 'POST',
    payload,
    accessToken
  });
  return transaction;
};
