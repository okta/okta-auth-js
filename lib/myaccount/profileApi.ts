import { sendRequest } from './request';
import { IAPIFunction } from './types';
import {
  ProfileTransaction,
  ProfileSchemaTransaction
} from './transactions';

/**
 * @scope: okta.myAccount.profile.read
 */
export const getProfile: IAPIFunction<ProfileTransaction> = async (oktaAuth, options?) => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile',
    method: 'GET',
    accessToken: options?.accessToken,
  }, ProfileTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.profile.manage
 */
export const updateProfile: IAPIFunction<ProfileTransaction> = async (
  oktaAuth, 
  options
) => {
  const { payload, accessToken } = options!;
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile',
    method: 'PUT',
    payload,
    accessToken,
  }, ProfileTransaction);
  return transaction;
};

/**
 * @scope: okta.myAccount.profile.read
 */
export const getProfileSchema: IAPIFunction<ProfileSchemaTransaction> = async (
  oktaAuth, 
  options?
): Promise<ProfileSchemaTransaction> => {
  const transaction = await sendRequest(oktaAuth, {
    url: '/idp/myaccount/profile/schema',
    method: 'GET',
    accessToken: options?.accessToken,
  }, ProfileSchemaTransaction);
  return transaction;
};
