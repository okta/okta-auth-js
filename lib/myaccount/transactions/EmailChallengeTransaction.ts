import EmailStatusTransaction from './EmailStatusTransaction';
import { 
  EmailProfile, 
  Status,
  VerificationPayload, 
} from '../types';
import BaseTransaction from './Base';
import { generateRequestFnFromLinks } from '../request';

export default class EmailChallengeTransaction extends BaseTransaction {
  id: string;
  expiresAt: string;
  profile: EmailProfile;
  status: Status;

  poll: () => Promise<EmailStatusTransaction>;
  // eslint-disable-next-line no-use-before-define
  verify: (payload: VerificationPayload) => Promise<EmailChallengeTransaction>;

  constructor(oktaAuth, options) {
    super(oktaAuth, options);

    const { accessToken, res } = options;
    // assign required fields from res
    const { id, expiresAt, profile, status, _links } = res;
    this.id = id;
    this.expiresAt = expiresAt;
    this.profile = profile;
    this.status = status;

    // assign transformed fns to transaction
    this.poll = async () => {
      const fn = generateRequestFnFromLinks({ 
        oktaAuth, 
        accessToken, 
        methodName: 'poll', 
        links: _links,
      }, EmailStatusTransaction);
      return await fn();
    };
    this.verify = async (payload) => {
      const fn = generateRequestFnFromLinks({ 
        oktaAuth, 
        accessToken, 
        methodName: 'verify', 
        links: _links,
      }, EmailChallengeTransaction);
      return await fn(payload);
    };
  }
}
