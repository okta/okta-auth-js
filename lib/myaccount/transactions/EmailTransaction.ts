import EmailChallengeTransaction from './EmailChallengeTransaction';
import EmailStatusTransaction from './EmailStatusTransaction';
import { EmailProfile, EmailRole, Status, VerificationPayload } from '../types';
import BaseTransaction from './Base';
import { generateRequestFnFromLinks } from '../request';

export default class EmailTransaction extends BaseTransaction {
  id: string;
  profile: EmailProfile;
  roles: EmailRole[];
  status: Status;

  // eslint-disable-next-line no-use-before-define
  get: () => Promise<EmailTransaction>;
  delete: () => Promise<BaseTransaction>;
  challenge: () => Promise<EmailChallengeTransaction>;
  poll?: () => Promise<EmailStatusTransaction>;
  verify?: (payload: VerificationPayload) => Promise<BaseTransaction>;

  constructor(oktaAuth, options) {
    super(oktaAuth, options);

    const { accessToken, res } = options;
    // assign required fields from res
    const { id, profile, roles, status, _links } = res;
    this.id = id;
    this.profile = profile;
    this.roles = roles;
    this.status = status;

    // assign transformed fns to transaction
    this.get = async () => {
      const fn = generateRequestFnFromLinks({ 
        oktaAuth, 
        accessToken, 
        methodName: 'get', 
        links: _links,
      }, EmailTransaction);
      return await fn();
    };
    this.delete = async () => {
      const fn = generateRequestFnFromLinks({ 
        oktaAuth, 
        accessToken, 
        methodName: 'delete', 
        links: _links 
      });
      return await fn();
    };
    this.challenge = async () => {
      const fn = generateRequestFnFromLinks({ 
        oktaAuth, 
        accessToken, 
        methodName: 'challenge', 
        links: _links,
      }, EmailChallengeTransaction);
      return await fn();
    };
    if (_links.poll) {
      this.poll = async () => {
        const fn = generateRequestFnFromLinks({ 
          oktaAuth, 
          accessToken, 
          methodName: 'poll', 
          links: _links,
        }, EmailStatusTransaction);
        return await fn();
      };
    }
    if (_links.verify) {
      this.verify = async (payload: VerificationPayload) => {
        const fn = generateRequestFnFromLinks({ 
          oktaAuth, 
          accessToken, 
          methodName: 'verify', 
          links: _links,
        });
        return await fn(payload);
      };
    }
  }
}
