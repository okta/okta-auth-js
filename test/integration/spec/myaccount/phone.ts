import {
  addPhone,
  deletePhone,
  sendPhoneChallenge,
  verifyPhoneChallenge
} from '../../../../lib/myaccount';
import {
  createClient,
  signinAndGetTokens
} from '../../util';
import { PhoneTransaction } from '../../../../lib/myaccount/transactions';

// TODO: generate dynamic profile with a18n api
describe('MyAccount Phone API', () => {
  let client, token;

  beforeAll(async () => {
    client = createClient({});
    const {
      tokens: {
        accessToken: { accessToken } = {}
      }
    } = await signinAndGetTokens(client, {
      scopes: [
        'openid',
        'profile',
        'okta.myAccount.phone.read',
        'okta.myAccount.phone.manage'
      ],
      acrValues: 'urn:okta:loa:2fa:any:ifpossible',
    });
    token = accessToken!;
  });

  describe('Manage with Transaction functions', () => {
    let phone: PhoneTransaction;

    afterEach(async () => {
      if (phone) {
        try {
          await phone.delete();
        } catch {
          // do nothing
        }
      }
    });

    it.only('can manage phone with transaction functions', async () => {
      // create test phone
      phone = await addPhone(client, {
        accessToken: token,
        payload: {
          profile: {
            phoneNumber: '+11234567890'
          },
          // verify link will be included in the response when sendCode is true
          // keep this flag as false in test to avoid sms rate limit issue
          sendCode: false,
          method: 'SMS'
        }
      });
      expect(phone).toMatchSnapshot({
        headers: expect.any(Object),
        id: expect.any(String),
        profile: {
          phoneNumber: expect.any(String)
        }
      });

      // get phone
      const transaction = await phone.get();
      expect(transaction).toMatchSnapshot({
        headers: expect.any(Object),
        id: expect.any(String),
        profile: {
          phoneNumber: expect.any(String)
        }
      });

      // send challenge code
      // api only send challenge to real number
      // TODO: use a18n for testing when it's scalable
      // disable challenge test for now
      // const challenge = await phone.challenge({ method: 'SMS' });
      // expect(challenge).toMatchSnapshot();

      // verify challenge code
      // no verify link when auto send is false
      // use sdk method instead
      // await expect(phone.verify!({ verificationCode: '000000' }))
      //   .rejects
      //   .toThrowErrorMatchingSnapshot();
      await expect(verifyPhoneChallenge(client, {
        accessToken: token,
        id: phone.id,
        payload: { verificationCode: '000000' }
      }))
        .rejects
        .toThrowErrorMatchingSnapshot();

      // delete
      const deleteTransaction = await phone.delete();
      expect(deleteTransaction).toMatchSnapshot();
    });

  });

  describe('Manage with SDK methods', () => {
    let phone: PhoneTransaction;

    afterEach(async () => {
      if (phone) {
        try {
          await phone.delete();
        } catch {
          // do nothing
        }
      }
    });

    // TODO: integrate with i18n
    it('can manage phone with SDK methods', async () => {
      // create test phone
      phone = await addPhone(client, {
        accessToken: token,
        payload: {
          profile: {
            phoneNumber: '+12345678901'
          },
          sendCode: false,
          method: 'SMS'
        }
      });
      expect(phone).toMatchSnapshot({
        headers: expect.any(Object),
        id: expect.any(String),
        profile: {
          phoneNumber: expect.any(String)
        }
      });

      // send challenge code
      try {
        const challenge = await sendPhoneChallenge(client, {
          id: phone.id,
          payload: { method: 'SMS' },
          accessToken: token,
        });
        expect(challenge).toMatchSnapshot({
          headers: expect.any(Object),
        });
      } catch (err) {
        // may failure due to - You have reached the limit of sms requests, please try again later.
        console.log(err);
      }

      // verify challenge
      // use invalid code to test the api. expect error
      // TODO: integrate with a18n api to test against real verification code
      await expect(verifyPhoneChallenge(client, {
        accessToken: token,
        id: phone.id,
        payload: { verificationCode: '000000' }
      }))
        .rejects
        .toThrowErrorMatchingSnapshot();

      // delete
      const deleteTransaction = await deletePhone(client, {
        accessToken: token,
        id: phone.id
      });
      expect(deleteTransaction).toMatchSnapshot();
    });

  });

});
