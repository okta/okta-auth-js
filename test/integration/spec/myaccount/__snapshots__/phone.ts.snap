// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 1`] = `
Object {
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "phoneNumber": Any<String>,
  },
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 2`] = `
Object {
  "challenge": [Function],
  "delete": [Function],
  "get": [Function],
  "headers": Any<Object>,
  "id": Any<String>,
  "profile": Object {
    "phoneNumber": Any<String>,
  },
  "status": "UNVERIFIED",
}
`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 3`] = `"Api validation failed: factorId and/or verificationCode"`;

exports[`MyAccount Phone API Manage with Transaction functions can manage phone with transaction functions 4`] = `BaseTransaction {}`;
